Utils/src/HiddenSectorProducer.cc

```diff
@@ -22,143 +27,390 @@
+    void firstDark(CandPtr part, CandSet& firstMd, CandSet& firstQd, CandSet& firstGd, CandSet& firstQdM, CandSet& firstQsM, CandPtr& firstQdM1, CandPtr& firstQdM2, CandPtr& firstQsM1, CandPtr& firstQsM2, bool& secondDM, bool& secondSM) const;
+    int checkLast(const reco::GenJet& jet, CandSet stableDs, int value, double& frac) const;
+    int checkFirst(const reco::GenJet& jet, const CandSet& firstP, int value) const;
+    void matchPFMtoJet(CandPtr& part, std::vector<LorentzVector>& matchedJets, const reco::GenJet& jet, int& nPartPerJet, int& t_MT2JetID, const int& mt2ID) const;
+    std::vector<int> matchGenRec(const edm::View<pat::Jet>& jets, const edm::View<reco::GenJet>& gen) const;
+    edm::InputTag JetTag_, MetTag_, GenMetTag_, GenTag_, GenJetTag_;
+    edm::EDGetTokenT<edm::View<pat::Jet>> JetTok_;
+    edm::EDGetTokenT<edm::View<pat::MET>> MetTok_;
+    edm::EDGetTokenT<edm::View<reco::GenMET>> GenMetTok_;
+    edm::EDGetTokenT<edm::View<reco::GenParticle>> GenTok_;
+    edm::EDGetTokenT<edm::View<reco::GenJet>> GenJetTok_;
+    double coneSize_;
+    PidSet DarkMediatorIDs_, DarkQuarkIDs_, DarkHadronIDs_, DarkGluonIDs_, DarkStableIDs_, DarkFirstIDs_, SMQuarkIDs_;
 };
 
+void HiddenSectorProducer::fillSet(PidSet& IDset, const std::string& name, const edm::ParameterSet& iConfig){
+  const auto& ids = iConfig.getParameter<std::vector<unsigned>>(name);
+  IDset.insert(ids.begin(),ids.end());
+}
+
+bool HiddenSectorProducer::isDark(const PidSet& darkList, CandPtr part) const {
+  return isDark(darkList, part->pdgId());
+}
+
+bool HiddenSectorProducer::isDark(const PidSet& darkList, int pid) const {
+  return darkList.find(std::abs(pid)) != darkList.end();
+}
+
+bool HiddenSectorProducer::isDark(const CandSet& darkList, CandPtr part) const {
+  return darkList.find(part) != darkList.end();
+}
+
+bool HiddenSectorProducer::isAncestor(const PidSet& darkList, CandPtr part) const {
+  if(isDark(darkList, part)) return true;
+  for(size_t i=0;i< part->numberOfMothers();i++)
+  {
+    if(isAncestor(darkList,part->mother(i))) return true;
+  }
+  return false;
+}
+
+void HiddenSectorProducer::medDecay(CandPtr part, CandSet& firstQdM, CandSet& firstQsM, CandPtr& firstQdM1, CandPtr& firstQdM2, CandPtr& firstQsM1, CandPtr& firstQsM2, bool& secondDM, bool& secondSM) const {
+  for(unsigned i = 0; i < part->numberOfDaughters(); i++){
+    CandPtr dau = part->daughter(i);
+    // if the first dark mediator's daughter is still a dark mediator, then check the daughters of this daughter dark mediator until we get daughters that are not dark mediator
+    if(isDark(DarkMediatorIDs_,dau)) medDecay(dau,firstQdM,firstQsM,firstQdM1,firstQdM2,firstQsM1,firstQsM2,secondDM,secondSM);
+    else{
+      if(isDark(DarkQuarkIDs_,dau)){
+        firstQdM.insert(dau);
+        if(secondDM == false){
+          firstQdM1 = dau;
+          secondDM = true;
+        }
+        else firstQdM2 = dau;
+      }
+      else if(isDark(SMQuarkIDs_,dau)){
```

Comment:
now that we're using `isDark` with SM quarks also, maybe we should change the name to `isParticle` (otherwise it's confusing)